File: happyandunhappynumbers.py
from sys import stdin,stdout

def sum_square(n):
    t = 0
    while n > 0:
        d = n % 10
        t += d * d
        n //= 10
    return t

def check_happy(n):
    seen = set()
    while n > 9:
        n = sum_square(n)
        if n in seen:
            return 0
        seen.add(n)
    return 1 if n == 1 or n == 7 else 0

happy = [0 for _ in range(1000001)]
for i in range(1,1000001):
    happy[i] += happy[i - 1] + check_happy(i)

for _ in range(int(stdin.readline())):
    A,B = map(int,stdin.readline().split())
    stdout.write(f"{happy[B] - happy[A - 1]}\n")

File: happyandunhappynumbers.cpp
#include <bits/stdc++.h>
#include <unordered_set>

using namespace std;

typedef unordered_set<int> usi;
int happy[1000001];

int sumSquare(int n) {
    int t = 0;
    while (n > 0) {
        int d = n % 10;
        t += d * d;
        n /= 10;
    }
    return t;
}

int checkHappy(int n) {
    usi seen;
    while (n > 9) {
        n = sumSquare(n);
        if (seen.find(n) != seen.end()) return 0;
        seen.insert(n);
    }
    return (n == 1 || n == 7) ? 1 : 0;
}

int main() {
    //precompute
    happy[0] = 0;
    for (int i = 1; i < 1000001; i++) happy[i] = happy[i - 1] + checkHappy(i);
    int T; scanf("%d",&T);
    while (T--) {
        int A,B; scanf("%d %d",&A,&B);
        printf("%d\n",happy[B] - happy[A - 1]);
    }
    return 0;
}